designs/kindless-upgrades.md

```diff
@@ -0,0 +1,150 @@
+
+#### Separating etcd from CP changes
+
+Currently, the controller performs changes to the etcd cluster and CP all at once in one single step. This is no a problem from a workflow perspective, since the etcdadm cluster, CAPI cluster and KCP controllers orchestrate the operation in the right order. However, it makes recovering from an error more difficult since there are more changes to "rollback".
+
+Most of the time when an upgrade fails, it shows up in the first etcd node not coming up. By making the etcd changes a separate step in the upgrade process, we reduce the amount of changes to rollback in case of an issue. If the first etcd node doesn't come up, the etcd cluster spec can be manually changed to its previous version and the new machine deleted, leaving the cluster in the same state as before.
+
+This will require changing the way the etcdadm cluster, core CAPI cluster and KCP controllers interact, or at least make the current behavior optional. Right now, the system is designed to apply all the CP changes at once, so it orchestrates itself making that assumption. If this we don't change this, it will cause a double CP upgrade, one after etcd finishes the upgrade (when it updated the endpoints secret) and a second one after updating the KCP.
+
+#### Waiting for control plane to be fully upgraded
+
+The current implementation of the controller first upgrades the CP and the just waits for `ControlPlaneReady` condition in the cluster object before continuing with the rest of the reconciliation (CNI, workers, etc.).
+
+This is not enough. There can be delays, creating a race condition where the status of the CAPI cluster is not updated fast enough after updating the KCP. This is more common when using external etcd, since the KCP won't be `NotReady` until etcd is upgraded.
+
+This implementation needs to extended to look at the KCP instead of the CAPI cluster, making sure `generation` and `observedGeneration` are leveraged to avoid trusting outdated statuses.
+
+This could make the reconciliation process take longer for control planes with a lot of nodes, since the controller would wait until all new nodes have joined the cluster and all the old ones have been removed. In order to speed things up, we might not need to wait until the whole CP is upgraded before moving on with the next phases. It might be enough with just waiting for the first control plane node to be upgraded (so first new machine created and first old machine deleted).
+
+This gives us good confidence the CP upgrade is going well. That said, since the next phase involves upgrading the CNI at that can affect networking, we recommend starting with extra caution and just wait for the whole CP to be upgraded (all new machines created and ready, all old machines deleted).
+
+That said, the "first node wait" strategy can still be a great alternative for workload clusters, given their upgrade process is less risky.
+
+#### Leaving `kube-vip` CP node for last
```

Comment:
Is kube-vip the only concern or is the scheduling of management components on the newer nodes also something to think about? I think the main issue is api server connectivity even in the management components so maybe kube-vip is the only thing.

Reply:
Ah interesting, allow me to dig in here. What management components are you referring to? Like the EKS-A and CAPI controllers?

Assuming it's EKS-A and CAPI controllers, I didn't consider that a concern, but I might be missing some scenarios. I was thinking that if the logic to "continue" with the reconciliation is only present in the controllers, the reconciliation won't progress unless they get successfully scheduled in a new node.

Or were you saying that we might want to leave the node running the controllers to be deleted the last? That might be a good idea but it can become tricky, since they can probably be running in different nodes. I'm not sure what the logic should be here but something to consider. Thoughts?

Reply:
Ah got you. Yeah I'm not concerned about transient errors due to webhooks restarting, the control loop should requeue those requests and retry until the webhooks are available.